### PR TITLE
Spring data features added

### PR DESCRIPTION
**Pull request by RomanYuzviak**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                    nullable: false
+              - column:
+                  name: isbn
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+                    unique: true
+              - column:
+                  name: price
+                  type: numeric
```
Comment (AlexVulpine) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml:34]: Suggest adding a constraint nullable: false

File: src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public BookDto get(Long id) {
+        return bookMapper.toDto(bookRepository.findById(id).orElseThrow(() ->
+                new RuntimeException("There is not book in db by id %d"
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java]: Use your EntityNotFoundException here

File: src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Override
+    public void delete(Long id) {
+        bookRepository.deleteById(id);
+    }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto bookRequestDto) {
+        if (!bookRepository.existsById(id)) {
+            throw new RuntimeException("There is not book in db by id %d"
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java]: Use your EntityNotFoundException here also

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.example.onlinebookstore.mapper;
+
+import com.example.onlinebookstore.config.MapperConfig;
+import com.example.onlinebookstore.dto.BookDto;
+import com.example.onlinebookstore.dto.CreateBookRequestDto;
+import com.example.onlinebookstore.model.Book;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java:10]: I'd suggest putting the config info directly here, bc we won't need a very detailed config for mapper.

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+public interface BookMapper {
+    BookDto toDto(Book book);
+
+    Book toModel(CreateBookRequestDto bookDto);
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java]: It's better to change name to "toBook" to be more concise as we will have more classes in model package in the future.

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+@RestController
+@RequestMapping(value = "/api/books")
+public class BookController {
+    private final BookService bookService;
+
+    @Autowired
+    public BookController(BookService bookService) {
+        this.bookService = bookService;
+    }
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]: You could use RequiredArgsConstructor annotation from lombok to get rid of this.

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    @GetMapping
+    public List<BookDto> getAll() {
+        return bookService.findAll();
+    }
+
+    @GetMapping("/{id}")
+    public BookDto findById(@PathVariable Long id) {
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]: suggest naming it getById

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @PostMapping
+    public BookDto save(@RequestBody CreateBookRequestDto requestDto) {
+        return bookService.save(requestDto);
+    }
+
+    @PutMapping("/{id}")
+    BookDto update(@PathVariable Long id, @RequestBody CreateBookRequestDto requestDto) {
+        return bookService.update(id, requestDto);
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:46]: You also need @ResponseStatus(HttpStatus.ACCEPTED) here

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @GetMapping("/{id}")
+    public BookDto findById(@PathVariable Long id) {
+        return bookService.get(id);
+    }
+
+    @PostMapping
+    public BookDto save(@RequestBody CreateBookRequestDto requestDto) {
+        return bookService.save(requestDto);
+    }
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]: As we added CreateBookRequest for implementing Create endpoint suggest naming it createBook. You also need to add annotation with ResponseStatus

File: src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java [hunk 10]
```diff
@@ -1,13 +1,16 @@
...
 
+import com.example.onlinebookstore.service.BookService;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
 @SpringBootApplication
 public class OnlineBookStoreApplication {
+    @Autowired
+    private BookService bookService;
```
Comment (AlexVulpine) [inline src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java]: Better make it final

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
+
+
+
+
+
```
Comment (AlexVulpine) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]: Redundant lines

**Review by AlexVulpine — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: pom.xml [hunk 12]
```diff
@@ -16,13 +16,24 @@
     <properties>
         <java.version>17</java.version>
         <maven.checkstyle.plugin.configLocation>checkstyle.xml</maven.checkstyle.plugin.configLocation>
+        <mapstruct.version>1.5.5.Final</mapstruct.version>
+        <lombok.mapstruct.binding.version>0.2.0</lombok.mapstruct.binding.version>
+        <lombok.version>1.18.30</lombok.version>
```
Comment (NazarSmith228) [inline pom.xml]: this is redundant - lombok is  inherited from `spring-boot-starter-parent`

File: pom.xml [hunk 13]
```diff
@@ -38,6 +49,25 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-maven-plugin</artifactId>
+            <version>${liquibase.version}</version>
```
Comment (NazarSmith228) [inline pom.xml]: this is a plugin

therefore it should be removed from dependencies section

File: pom.xml [hunk 14]
```diff
@@ -38,6 +49,25 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.mapstruct</groupId>
+            <artifactId>mapstruct</artifactId>
+            <version>${mapstruct.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>${liquibase.version}</version>
```
Comment (NazarSmith228) [inline pom.xml]: no need to specify the version here - it is taken from `spring-boot-starter-parent`

File: src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java (hunk 10 above)
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java]: why do you need it here?

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        return bookService.get(id);
+    }
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    public BookDto save(@RequestBody CreateBookRequestDto requestDto) {
+        return bookService.save(requestDto);
+    }
+
+    @ResponseStatus(HttpStatus.ACCEPTED)
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]: HTTP 202 ACCEPTED is wrong here

HTTP 202 means that a request was accepted for processing but hasn't been processed yet - the processing should happen in the nearest future

in our case update is executed immediately, thus we need to use HTTP 200 OK

File: src/main/java/com/example/onlinebookstore/dto/BookDto.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
+package com.example.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class BookDto {
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/dto/BookDto.java]: please use records for DTO

if there is a problem with Checkstyle - upgrade the checkstyle-plugin to 3.3.0

File: src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package com.example.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class CreateBookRequestDto {
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java]: please use records for DTO

if there is a problem with Checkstyle - upgrade the checkstyle-plugin to 3.3.0

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    private String title;
+    @Column(nullable = false)
+    private String author;
+    @Column(nullable = false, unique = true)
+    private String isbn;
+    @Column(nullable = false)
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
+    @Column(nullable = false)
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/model/Book.java:33]: shouldn't you specify `name = "is_deleted" here?

will it work as it is?

File: src/main/java/com/example/onlinebookstore/repository/BookRepository.java [hunk 19]
```diff
@@ -0,0 +1,8 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface BookRepository extends JpaRepository<Book, Long> {
+
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/repository/BookRepository.java]: redundant blank line

File: src/main/java/com/example/onlinebookstore/repository/BookRepository.java [hunk 20]
```diff
@@ -0,0 +1,8 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface BookRepository extends JpaRepository<Book, Long> {
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/repository/BookRepository.java:8]: `@Repository` is missing

File: src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public BookDto get(Long id) {
+        return bookMapper.toDto(bookRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException("There is not book in db by id %d"
+                .formatted(id))));
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java]: you may simplify the setup here a bit:
```
return repo.findById(id)
       .map(mapper::toDto)
       .orElseThrow(...);
```

File: src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto bookRequestDto) {
+        if (!bookRepository.existsById(id)) {
+            throw new EntityNotFoundException("There is not book in db by id %d"
+                    .formatted(id));
+        }
+        Book updatedBook = (bookMapper.toBook(bookRequestDto));
+        updatedBook.setId(id);
+        return bookMapper.toDto(bookRepository.save(updatedBook));
```
Comment (NazarSmith228) [inline src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java]: bad approach

you should fetch the desired book by id, then update it, and save afterwards

you can leverage Mapstruct to perform update:
```
void updateBook(CreateBookRequestDto  dto, @MappingTarget Book book);
```

the final version will be like:
```
Book book = repo.findById(id).orElseThrow(...);
mapper.updateBook(dto, book);
return mapper.toDto(repo.save(book));
```

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/com/example/onlinebookstore/model/Book.java (hunk 18 above)
Comment (RomanYuzviak) [reply src/main/java/com/example/onlinebookstore/model/Book.java:33]: Yes, It works

File: src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java [hunk 23]
```diff
@@ -1,10 +1,12 @@
 package com.example.onlinebookstore;
 
+import com.example.onlinebookstore.service.BookService;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
 @SpringBootApplication
 public class OnlineBookStoreApplication {
+    private BookService bookService;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java]: Remove unused service

**Review by Elena-Bruyako — APPROVED** [review]
Good job! Left one comment.

**Pull request merged**